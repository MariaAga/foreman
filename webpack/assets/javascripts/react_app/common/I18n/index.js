import intl from './intl.js';
import { jed, translate, ngettext, sprintf } from './jed.js';

export const documentLocale = async () => {
  await new Promise(resolve => setImmediate(resolve));
  return intl.locale;
};

export { translate, ngettext, jed, sprintf, intl };

export default { translate, ngettext, jed, sprintf, intl };

window.__ = translate;
window.n__ = ngettext;
