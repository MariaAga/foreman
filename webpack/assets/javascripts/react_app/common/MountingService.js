import ReactDOM from 'react-dom';
import React, { useState, useEffect } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';

export { default as registerReducer } from '../redux/reducers/registerReducer';

// eslint-disable-next-line react/prop-types
const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const [allPluginsImported, setAllPluginsImported] = useState(false);
  async function mountComponent() {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else if (allPluginsImported) {
      const awaitedComponent = componentRegistry.markup(component, {
        data,
        store,
        flattenData,
      });
      setMounted(true);
      setMountedComponent(awaitedComponent);
    }
  }
  const updateAllPluginsImported = e => {
    if (Object.values(window.allPluginsLoaded).every(Boolean))
      setAllPluginsImported(true);
  };
  useEffect(() => {
    document.addEventListener('loadPlugin', updateAllPluginsImported);
    return () =>
      window.removeEventListener('loadPlugin', updateAllPluginsImported);
  }, []);
  useEffect(() => {
    if (!mounted) mountComponent();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [allPluginsImported]);
  return mounted ? mountedComponent : <div>Loading...</div>;
};

function mountNode(component, reactNode, data, flattenData) {
  ReactDOM.render(
    <AwaitedMount
      component={component}
      data={data}
      flattenData={flattenData}
    />,
    reactNode
  );
}

/**
 * This is a html tag (Web component) that can be used for mounting react component from ComponentRegistry.
 */
class ReactComponentElement extends HTMLElement {
  static get observedAttributes() {
    return ['data-props'];
  }

  get componentName() {
    return this.getAttribute('name');
  }
  get reactProps() {
    return this.dataset.props !== '' ? JSON.parse(this.dataset.props) : {};
  }
  set reactProps(newProps) {
    this.dataset.props = JSON.stringify(newProps);
  }
  get mountPoint() {
    if (!this._mountPoint) {
      this._mountPoint = this;
    }

    return this._mountPoint;
  }

  attributeChangedCallback(name, oldValue, newValue) {
    switch (name) {
      case 'data-props':
        // if this is not the initial prop set
        if (oldValue !== null) this._render();
        break;
      default:
      // We don't know how to react to default attribute change
    }
  }

  connectedCallback() {
    this._render();
  }

  _render() {
    try {
      mountNode(this.componentName, this, this.reactProps, true);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(
        `Unable to mount foreman-react-component: ${this.componentName}`,
        error
      );
    }
  }
}

if (!window.customElements.get('foreman-react-component')) {
  window.customElements.define(
    'foreman-react-component',
    ReactComponentElement
  );
}
