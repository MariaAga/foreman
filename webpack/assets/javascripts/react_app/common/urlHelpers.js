/**
 * Build a url from given controller, action and id
 * @param {String} controller - the controller
 * @param {String} action - the action
 */
export const urlBuilder = (controller, action, id = undefined) =>
  `/${controller}/${id ? `${id}/` : ''}${action}`;

/**
 * Build a url with search query
 * @param {String} base - the base url
 * @param {String} searchQuery - the search query
 */
export const urlWithSearch = (base, searchQuery) =>
  `/${base}?search=${searchQuery}`;
