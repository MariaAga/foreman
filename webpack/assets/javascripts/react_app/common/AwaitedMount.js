import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';
import { translate as __ } from './I18n';

export const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const [allPluginsImported, setAllPluginsImported] = useState(false);
  async function mountComponent() {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else if (allPluginsImported) {
      const awaitedComponent = componentRegistry.markup(component, {
        data,
        store,
        flattenData,
      });
      setMounted(true);
      setMountedComponent(awaitedComponent);
    }
  }
  const updateAllPluginsImported = e => {
    setAllPluginsImported(true);
  };
  useEffect(() => {
    document.addEventListener('loadJS', updateAllPluginsImported);
    return () => window.removeEventListener('loadJS', updateAllPluginsImported);
  }, []);
  useEffect(() => {
    if (!mounted) mountComponent();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [allPluginsImported]);
  return mounted ? mountedComponent : <div>{__('Loading...')}</div>;
};

AwaitedMount.propTypes = {
  component: PropTypes.string.isRequired,
  data: PropTypes.object,
  flattenData: PropTypes.bool,
};
AwaitedMount.defaultProps = {
  data: {},
  flattenData: false,
};
