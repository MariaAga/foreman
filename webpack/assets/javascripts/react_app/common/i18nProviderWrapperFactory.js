import React from 'react';
import { IntlProvider } from 'react-intl';
import { intl } from './I18n';
import { getDisplayName } from './helpers';

const i18nProviderWrapperFactory = () => WrappedComponent => {
  const wrappedName = getDisplayName(WrappedComponent);
  const I18nProviderWrapper = (props) => {
    const [htmlElemnt] = document.getElementsByTagName('html');
    const langAttr = htmlElemnt.getAttribute('lang') || intl.locale;
    const timezone = htmlElemnt.getAttribute('data-timezone') || intl.timezone;
    return (
      <IntlProvider
        locale={langAttr}
        timeZone={timezone}
      >
          <WrappedComponent {...props} />
      </IntlProvider>
    );
  };

  I18nProviderWrapper.displayName = `I18nProviderWrapper(${wrappedName})`;

  return I18nProviderWrapper;
};

export { i18nProviderWrapperFactory };
