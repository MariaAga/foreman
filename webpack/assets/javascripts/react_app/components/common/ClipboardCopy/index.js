import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button } from 'patternfly-react';
import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import { translate as __ } from '../../../common/I18n';
import './clipboard-copy.scss';

const ClipboardCopy = ({
  text: defaultText,
  successMessage,
  buttonText,
  textareaProps,
  buttonProps,
}) => {
  const [text, setText] = useState(defaultText);

  return (
    <div className="clipboard-copy">
      <textarea
        defaultValue={text}
        onChange={({ target: { value } }) => setText(value)}
        {...textareaProps}
      />
      <Tooltip
        content={successMessage}
        position={TooltipPosition.right}
        trigger="click"
      >
        <Button
          onClick={() => navigator.clipboard.writeText(text)}
          bsStyle="default"
          {...buttonProps}
        >
          {buttonText}
        </Button>
      </Tooltip>
    </div>
  );
};

ClipboardCopy.propTypes = {
  text: PropTypes.string.isRequired,
  buttonText: PropTypes.string,
  successMessage: PropTypes.string,
  textareaProps: PropTypes.object,
  buttonProps: PropTypes.object,
};

ClipboardCopy.defaultProps = {
  buttonText: __('Copy to clipboard'),
  successMessage: __('Copied!'),
  textareaProps: {},
  buttonProps: {},
};

export default ClipboardCopy;
