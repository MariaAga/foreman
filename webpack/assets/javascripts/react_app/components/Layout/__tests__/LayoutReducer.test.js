import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_IS_NAV_OPEN,
} from '../LayoutConstants';

import reducer from '../LayoutReducer';

import { testReducerSnapshotWithFixtures } from '../../../common/testHelpers';

const fixtures = {
  'should return the initial state': {},
  'should handle LAYOUT_INITIALIZE': {
    action: {
      type: LAYOUT_INITIALIZE,
      payload: {
        items: 'some-items',
        organization: 'some organization',
        location: 'some location',
      },
    },
  },
  'should handle LAYOUT_SHOW_LOADING': {
    action: {
      type: LAYOUT_SHOW_LOADING,
    },
  },
  'should handle LAYOUT_HIDE_LOADING': {
    action: {
      type: LAYOUT_HIDE_LOADING,
    },
  },
  'should handle LAYOUT_CHANGE_IS_NAV_OPEN': {
    action: {
      type: LAYOUT_CHANGE_IS_NAV_OPEN,
      payload: { isNavOpen: false },
    },
  },
};

describe('Layout reducer', () =>
  testReducerSnapshotWithFixtures(reducer, fixtures));
