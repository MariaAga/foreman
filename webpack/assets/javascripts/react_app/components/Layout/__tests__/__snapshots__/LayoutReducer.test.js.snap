// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout reducer should handle LAYOUT_CHANGE_IS_NAV_OPEN 1`] = `
Object {
  "isLoading": false,
  "isNavOpen": false,
  "items": Array [],
}
`;

exports[`Layout reducer should handle LAYOUT_HIDE_LOADING 1`] = `
Object {
  "isLoading": false,
  "isNavOpen": true,
  "items": Array [],
}
`;

exports[`Layout reducer should handle LAYOUT_INITIALIZE 1`] = `
Object {
  "currentLocation": "some location",
  "currentOrganization": "some organization",
  "isLoading": false,
  "isNavOpen": undefined,
  "items": "some-items",
}
`;

exports[`Layout reducer should handle LAYOUT_SHOW_LOADING 1`] = `
Object {
  "isLoading": true,
  "isNavOpen": true,
  "items": Array [],
}
`;

exports[`Layout reducer should return the initial state 1`] = `
Object {
  "isLoading": false,
  "isNavOpen": true,
  "items": Array [],
}
`;
